### PR TITLE
En 4808/end of epoch trigger for shards

### PR DESCRIPTION
**Pull request by sasurobert**
added trigger for shard. Unit tests and integration test for happy case.

### CODE REVIEW

File: endOfEpoch/shardchain/trigger.go [hunk 1]
```diff
@@ -1,51 +1,334 @@
...
+	return t.getHeaderWithNonceAndHash(nonce, neededHash)
 }
 
 // Update updates the end-of-epoch trigger
 func (t *trigger) Update(round int64) {
 }
 
-// Processed signals that end-of-epoch has been processed
+// Processed sets end of epoch to false and cleans underlying structure
 func (t *trigger) Processed() {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply endOfEpoch/shardchain/trigger.go]: done

File: endOfEpoch/mock/chainStorerMock.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
+package mock
+
+import (
+	"github.com/ElrondNetwork/elrond-go/dataRetriever"
+	"github.com/ElrondNetwork/elrond-go/storage"
+	"github.com/pkg/errors"
+)
+
+// ChainStorerMock is a mock implementation of the ChianStorer interface
+type ChainStorerMock struct {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply endOfEpoch/mock/chainStorerMock.go]: done

File: endOfEpoch/mock/requestHandlerMock.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
+package mock
+
+type RequestHandlerMock struct {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply endOfEpoch/mock/requestHandlerMock.go]: done

File: endOfEpoch/mock/shardIdHashMapMock.go [hunk 4]
```diff
@@ -0,0 +1,32 @@
+package mock
+
+type ShardIdHasMapMock struct {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply endOfEpoch/mock/shardIdHashMapMock.go]: this is mock, as it is not implemented.

File: endOfEpoch/mock/uint64ByteSliceConverterMock.go [hunk 5]
```diff
@@ -0,0 +1,31 @@
+package mock
+
+// Uint64ByteSliceConverterMock converts byte slice to/from uint64
+type Uint64ByteSliceConverterMock struct {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply endOfEpoch/mock/uint64ByteSliceConverterMock.go]: done

File: endOfEpoch/mock/uint64SyncMapCacherStub.go [hunk 6]
```diff
@@ -0,0 +1,46 @@
+package mock
+
+import (
+	"github.com/ElrondNetwork/elrond-go/dataRetriever"
+)
+
+type Uint64SyncMapCacherStub struct {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply endOfEpoch/mock/uint64SyncMapCacherStub.go]: done

File: endOfEpoch/shardchain/trigger.go [hunk 7]
```diff
@@ -1,51 +1,334 @@
 package shardchain
 
 import (
+	"bytes"
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply endOfEpoch/shardchain/trigger.go]: done

File: process/interface.go [hunk 8]
```diff
@@ -1,6 +1,7 @@
 package process
 
 import (
+	"github.com/ElrondNetwork/elrond-go/sharding"
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/interface.go]: done

File: process/interface.go [hunk 9]
```diff
@@ -336,8 +364,12 @@ type VirtualMachinesContainerFactory interface {
 
 // EndOfEpochTriggerHandler defines that actions which are needed by processor for end of epoch
 type EndOfEpochTriggerHandler interface {
+	ReceivedHeader(header data.HeaderHandler)
 	IsEndOfEpoch() bool
 	Epoch() uint32
+	EpochStartRound() uint64
+	Processed()
+	Revert()
 	IsInterfaceNil() bool
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/interface.go]: done

File: integrationTests/testProcessorNode.go [hunk 10]
```diff
@@ -4,6 +4,8 @@ import (
 	"context"
 	"encoding/hex"
 	"fmt"
+	"github.com/ElrondNetwork/elrond-go/endOfEpoch/metachain"
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply integrationTests/testProcessorNode.go]: done

File: endOfEpoch/shardchain/trigger.go [hunk 11]
```diff
@@ -1,51 +1,334 @@
...
+			go t.requestHandler.RequestHeaderByNonce(sharding.MetachainShardId, i)
+		}
+	} else {
+		isMetaHdrFinal = true
+	}
+
+	return isMetaHdrFinal && isMetaHdrValid
+}
+
+func (t *trigger) getHeaderWithNonceAndHash(nonce uint64, neededHash []byte) (*block.MetaBlock, error) {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply endOfEpoch/shardchain/trigger.go]: yes. When it is refactored from process we might use it here as well. I did not want to use common functions from process.

File: integrationTests/testSyncNode.go [hunk 12]
```diff
@@ -45,6 +45,8 @@ func NewTestSyncNode(
 	tpn.initDataPools()
 	tpn.initTestNodeWithSync()
 
+	tpn.EndOfEpochTrigger = &mock.EndOfEpochTriggerStub{}
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply integrationTests/testSyncNode.go]: done.

File: endOfEpoch/shardchain/trigger.go [hunk 13]
```diff
@@ -1,51 +1,334 @@
...
+	DataPool       dataRetriever.PoolsHolder
+	Storage        dataRetriever.StorageService
+	RequestHandler endOfEpoch.RequestHandler
+
+	Epoch    uint32
+	Validity uint64
+	Finality uint64
 }
 
 type trigger struct {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply endOfEpoch/shardchain/trigger.go]: done a little.

File: integrationTests/multiShard/endOfEpoch/executingEpochChange_test.go [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+
+		for _, testNode := range nodes {
+			if testNode.ShardCoordinator.SelfId() == currentShId {
+				testHeader := testNode.BlockChain.GetCurrentBlockHeader()
+				assert.Equal(t, testHeader.GetNonce(), currentHeader.GetNonce())
+			}
+		}
+	}
+
+	//////////------ verify last added shardheaders in meta are with new epoch
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply integrationTests/multiShard/endOfEpoch/executingEpochChange_test.go]: done.

File: integrationTests/testProcessorNode.go [hunk 15]
```diff
@@ -631,11 +641,26 @@ func (tpn *TestProcessorNode) initBlockProcessor() {
 		Core:                         nil,
 		BlockChainHook:               tpn.BlockChainHookImpl,
 		ValidatorStatisticsProcessor: &mock.ValidatorStatisticsProcessorMock{},
-		EndOfEpochTrigger:            &mock.EndOfEpochTriggerStub{},
+		HeaderValidator:              headerValidator,
 		Rounder:                      &mock.RounderMock{},
 	}
 
 	if tpn.ShardCoordinator.SelfId() == sharding.MetachainShardId {
+		rounder := &mock.RounderMock{}
```
Comment (camilbancioiu) [inline integrationTests/testProcessorNode.go]: Why not reuse the `Rounder` from `argumentsBase`?

File: cmd/node/factory/structs.go [hunk 16]
```diff
@@ -634,9 +639,75 @@ func prepareGenesisBlock(args *processComponentsFactoryArgs, genesisBlocks map[u
 	return nil
 }
 
-func newEndOfEpochTrigger(args *processComponentsFactoryArgs, rounder endOfEpoch.Rounder) (endOfEpoch.TriggerHandler, error) {
+func newRequestHandler(
+	resolversFinder dataRetriever.ResolversFinder,
+	shardCoordinator sharding.Coordinator,
+) (process.RequestHandler, error) {
+	if shardCoordinator.SelfId() < shardCoordinator.NumberOfShards() {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply cmd/node/factory/structs.go]: in order to catch the error if it is a completely invalid shard

File: endOfEpoch/mock/headerValidatorStub.go [hunk 17]
```diff
@@ -0,0 +1,16 @@
+package mock
+
+import "github.com/ElrondNetwork/elrond-go/data"
+
+type HeaderValidatorStub struct {
+	IsHeaderConstructionValidCalled func(currHdr, prevHdr data.HeaderHandler) error
+}
+
+func (hvs *HeaderValidatorStub) IsHeaderConstructionValid(currHdr, prevHdr data.HeaderHandler) error {
+	return hvs.IsHeaderConstructionValidCalled(currHdr, prevHdr)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply endOfEpoch/mock/headerValidatorStub.go]: done

File: endOfEpoch/mock/poolsHolderStub.go [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+	TransactionsCalled         func() dataRetriever.ShardedDataCacherNotifier
+	UnsignedTransactionsCalled func() dataRetriever.ShardedDataCacherNotifier
+	RewardTransactionsCalled   func() dataRetriever.ShardedDataCacherNotifier
+	MiniBlocksCalled           func() storage.Cacher
+	MetaBlocksCalled           func() storage.Cacher
+	CurrBlockTxsCalled         func() dataRetriever.TransactionCacher
+}
+
+func (phs *PoolsHolderStub) CurrentBlockTxs() dataRetriever.TransactionCacher {
+	return phs.CurrBlockTxsCalled()
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply endOfEpoch/mock/poolsHolderStub.go]: not used, when used it is filled.

File: endOfEpoch/mock/uint64SyncMapCacherStub.go [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+	ClearCalled           func()
+	GetCalled             func(nonce uint64) (dataRetriever.ShardIdHashMap, bool)
+	MergeCalled           func(nonce uint64, src dataRetriever.ShardIdHashMap)
+	RemoveCalled          func(nonce uint64, shardId uint32)
+	RegisterHandlerCalled func(handler func(nonce uint64, shardId uint32, value []byte))
+	HasCalled             func(nonce uint64, shardId uint32) bool
+}
+
+func (usmcs *Uint64SyncMapCacherStub) Clear() {
+	usmcs.ClearCalled()
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply endOfEpoch/mock/uint64SyncMapCacherStub.go]: not used, when used it is filled.

File: endOfEpoch/shardchain/trigger.go [hunk 20]
```diff
@@ -1,51 +1,334 @@
...
+	}
+
+	t.mutReceived.Lock()
+
+	if len(metaHdr.EndOfEpoch.LastFinalizedHeaders) > 0 {
+		t.newEpochHdrReceived = true
+		t.mapEndOfEpochHdrs[string(hdrHash)] = metaHdr
+	} else {
+		t.mapHashHdr[string(hdrHash)] = metaHdr
+		t.mapNonceHashes[metaHdr.Nonce] = append(t.mapNonceHashes[metaHdr.Nonce], string(hdrHash))
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply endOfEpoch/shardchain/trigger.go]: done

File: endOfEpoch/shardchain/trigger.go [hunk 21]
```diff
@@ -1,51 +1,334 @@
...
+	t.mutReceived.Lock()
+	defer t.mutReceived.Unlock()
+
+	return t.isEndOfEpoch
 }
 
 // Epoch returns the current epoch number
 func (t *trigger) Epoch() uint32 {
-	return 0
+	return t.epoch
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply endOfEpoch/shardchain/trigger.go]: done

File: endOfEpoch/shardchain/trigger.go [hunk 22]
```diff
@@ -1,51 +1,334 @@
...
+		neededHdr, err := t.getHeaderWithNonceAndHash(i, currHdr.PrevHash)
+		if err != nil {
+			isMetaHdrValid = false
+		}
+
+		err = t.headerValidator.IsHeaderConstructionValid(currHdr, neededHdr)
+		if err != nil {
+			isMetaHdrValid = false
+		}
+	}
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply endOfEpoch/shardchain/trigger.go]: done

File: endOfEpoch/shardchain/trigger.go [hunk 23]
```diff
@@ -1,51 +1,334 @@
...
+
+	isMetaHdrFinal := false
+	nextBlocksVerified := uint64(0)
+	currHdr = metaHdr
+	for i := metaHdr.Nonce + 1; i <= metaHdr.Nonce+t.finality; i++ {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply endOfEpoch/shardchain/trigger.go]: done

File: endOfEpoch/shardchain/trigger.go [hunk 24]
```diff
@@ -1,51 +1,334 @@
...
+		if err != nil {
+			continue
+		}
+
+		currHdr = neededHdr
+		nextBlocksVerified += 1
+	}
+
+	if nextBlocksVerified < t.finality {
+		for i := currHdr.Nonce + 1; i <= currHdr.Nonce+t.finality; i++ {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply endOfEpoch/shardchain/trigger.go]: done

File: endOfEpoch/shardchain/trigger_test.go [hunk 25]
```diff
@@ -0,0 +1,263 @@
...
+	"github.com/ElrondNetwork/elrond-go/data"
+	"github.com/ElrondNetwork/elrond-go/data/block"
+	"github.com/ElrondNetwork/elrond-go/dataRetriever"
+	"github.com/ElrondNetwork/elrond-go/endOfEpoch"
+	"github.com/ElrondNetwork/elrond-go/endOfEpoch/mock"
+	"github.com/ElrondNetwork/elrond-go/storage"
+	"github.com/stretchr/testify/assert"
+)
+
+func creteMockShardEndOfEpochTriggerArguments() *ArgsShardEndOfEpochTrigger {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply endOfEpoch/shardchain/trigger_test.go]: done

File: integrationTests/testProcessorNode.go (hunk 15 above)
Comment (sasurobert) [reply integrationTests/testProcessorNode.go]: done

File: endOfEpoch/shardchain/trigger.go [hunk 26]
```diff
@@ -1,51 +1,334 @@
...
+	go t.requestHandler.RequestHeader(sharding.MetachainShardId, neededHash)
+
+	return nil, endOfEpoch.ErrMetaHdrNotFound
+}
+
+func (t *trigger) getHeaderWithNonceAndPrevHash(nonce uint64, prevHash []byte) (*block.MetaBlock, error) {
+	metaHdrHashesWithNonce := t.mapNonceHashes[nonce]
+	for _, hash := range metaHdrHashesWithNonce {
+		hdrWithNonce := t.mapHashHdr[hash]
+		if hdrWithNonce != nil && bytes.Equal(hdrWithNonce.PrevHash, prevHash) {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply endOfEpoch/shardchain/trigger.go]: no need. as it is ensured that adding to the map is correct.

File: integrationTests/multiShard/endOfEpoch/executingEpochChange_test.go [hunk 27]
```diff
@@ -94,4 +98,36 @@ func TestEndOfEpochChangeWithoutTransactionInMultiShardedEnvironment(t *testing.
...
+		if node.ShardCoordinator.SelfId() != sharding.MetachainShardId {
+			continue
+		}
+
+		currentMetaHdr, ok := node.BlockChain.GetCurrentBlockHeader().(*block.MetaBlock)
+		if !ok {
+			continue
+		}
+
+		shardHDrStorage := node.Storage.GetStorer(dataRetriever.BlockHeaderUnit)
```
Comment (iulianpascalau) [inline integrationTests/multiShard/endOfEpoch/executingEpochChange_test.go]: We should definitely build a service layer for these "getters" from blockchain/datapool/storage.

**Pull request merged**